binary-codec/types/type_utils.go

```diff
@@ -0,0 +1,169 @@
+package types
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+const (
+	MinIOUExponent  = -96
+	MaxIOUExponent  = 80
+	MaxIOUPrecision = 16
+	MinIOUMantissa  = 1000000000000000
+	MaxIOUMantissa  = 9999999999999999
+
+	NotXRPBitMask            = 0x80
+	PosSignBitMask           = 0x4000000000000000
+	ZeroCurrencyAmountHex    = 0x8000000000000000
+	NativeAmountByteLength   = 8
+	CurrencyAmountByteLength = 48
+
+	MinXRP   = 1e-6
+	MaxDrops = 1e17 // 100 billion XRP in drops aka 10^17
+
+	AllowedIOUValueCharacters = "0123456789.-eE"                                                                   // going to do this with Regex
+	AllowedIOUCodeCharacters  = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz?!@#$%^&*<>(){}[]|" // going to do this with Regex
+)
+
+type BigDecimal struct {
+	Scale         int
+	Precision     int
+	UnscaledValue string
+	Sign          int // 1 for negative, 0 for positive
+}
+
+type InvalidCharacterError struct {
+	AllowedChars string
+}
+
+func (e *InvalidCharacterError) Error() string {
+	return fmt.Sprintf("value contains invalid characters. Only the following are allowed: '%s'", e.AllowedChars)
+}
+
+type InvalidNativeCharacterError struct {
+	InvalidCharacter string
+}
+
+func (e *InvalidNativeCharacterError) Error() string {
+	return fmt.Sprintf("value contains invalid character '%s'", e.InvalidCharacter)
+}
+
+type InvalidAmountError struct {
+	Amount string
+}
+
+func (e *InvalidAmountError) Error() string {
+	return fmt.Sprintf("value '%s' is an invalid amount", e.Amount)
+}
+
+type OutOfRangeError struct {
+	Type string
+}
+
+func (e *OutOfRangeError) Error() string {
+	return fmt.Sprintf("%s is out of range", e.Type)
+}
+
+type InvalidCodeError struct {
+	Disallowed string
+}
+
+func (e *InvalidCodeError) Error() string {
+	return fmt.Sprintf("'%s' is/are disallowed or invalid", e.Disallowed)
+}
+
+// Creates a new custom BigDecimal object from a value string
+func NewBigDecimal(value string) (*BigDecimal, error) {
```

Comment:
lets move this and the struct into a pkg/bigDecimal package. Cause can get used outside of binary codec